spec/system/collections/user_views_collections_spec.rb

```diff
@@ -0,0 +1,18 @@
+require "rails_helper"
+
+RSpec.describe "Visiting collections", type: :system do
+  let(:user) { create(:user) }
+  let(:collection1) { create(:collection, :with_articles, user: user) }
+  let(:collection2) { create(:collection, user: user) }
+
+  before do
+    sign_in user
+    visit user_series_path(user.username)
+  end
+
+  it "shows all collections" do
+    [collection1, collection2].each do |collection|
+      expect(page.body).to have_link("#{collection.slug} (#{collection.articles.published.size} Part Series)")
```

Comment:
This expectation fails sometimes and I can't really get why 😖 I tried to debug this using `byebug`, but when I check for `user.collections.count` it returns `2` and the test never fails in that case, so I don't know how to proceed.

Reply:
Do you have an example message from one of the failed runs?

Reply:
One of the builds actually failed for that, so you can [see the message here](https://travis-ci.com/github/forem/forem/jobs/364361396#L2148-L2167).

Reply:
Does `collection2` also need articles? You have `:with_articles` on `collection1` but not `collection2`, so its size is probably 0. That's just a guess though, I didn't look into anything in detail.

Reply:
Hmmm, having it without articles was intentional, as the link should be present (and correct) for a collection without articles too.

Reply:
@AlbertoPdRF I had some time to look into it this morning. Here's how I debugged it in case that's useful to you in the future:

1. Add `:aggregate_failures` to the spec so we can see the results for each run.
    ```diff
    - it "shows all collections" do
    + it "shows all collections", :aggregate_failures do
    ```
    This revealed that both assertions are failing, not only the one for `collection2`.
2. Look at the page by adding `save_and_open_page` at the end of the spec. This showed the following:
    <img width="303" alt="Screen Shot 2020-07-27 at 10 25 30" src="https://user-images.githubusercontent.com/47985/88500670-8bf03800-cff3-11ea-8b0f-b5504f2c53f5.png">
3. Grepped for the message, found it in `app/views/collections/index.html.erb`. It will be rendered when `@collections.present?` is `false`.
4. Set a `binding.pry` in `app/controllers/collections_controller.rb`. `user.collections` was an empty array.
5. Changed the spec to use `let!` over `let`, green now. See e.g. https://mixandgo.com/learn/let-vs-instance-variables-in-rspec

If you want to try it out, I made a patch (with `git format-patch`) but for whatever reason GH won't let me attach it:

```
From d3d169d9e38e28066802fdc21a1faf1ae95e98d8 Mon Sep 17 00:00:00 2001
From: Michael Kohl <citizen428@dev.to>
Date: Mon, 27 Jul 2020 10:38:55 +0700
Subject: [PATCH 8/8] Fix spec

---
 spec/system/collections/user_views_collections_spec.rb | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/spec/system/collections/user_views_collections_spec.rb b/spec/system/collections/user_views_collections_spec.rb
index c664a0892..e078e5409 100644
--- a/spec/system/collections/user_views_collections_spec.rb
+++ b/spec/system/collections/user_views_collections_spec.rb
@@ -2,15 +2,15 @@ require "rails_helper"

 RSpec.describe "Visiting collections", type: :system do
   let(:user) { create(:user) }
-  let(:collection1) { create(:collection, :with_articles, user: user) }
-  let(:collection2) { create(:collection, user: user) }
+  let!(:collection1) { create(:collection, :with_articles, user: user) }
+  let!(:collection2) { create(:collection, user: user) }

   before do
     sign_in user
     visit user_series_path(user.username)
   end

-  it "shows all collections" do
+  it "shows all collections", :aggregate_failures do
     [collection1, collection2].each do |collection|
       expect(page.body).to have_link("#{collection.slug} (#{collection.articles.published.size} Part Series)")
     end
--
2.27.0
```

You can save this to your working directory and apply it with `git apply <filename>`.

Reply:
That's indeed very useful @citizen428! Thank you for looking into this and for the very clear explanation 😊